src/inc/engine/engine.h

```diff
@@ -15,13 +15,23 @@ class Engine
 {
 public:
 
-  Engine( std::vector<std::unique_ptr<GameComponent>>& components );
-
   ~Engine();
 
   void initialize( int height, int width );
 
-  void advance();
+  template< typename T, typename = typename std::enable_if_t<
+                          std::is_base_of<GameComponent, T>::value>>
+  void advance( std::vector<std::unique_ptr<T>>& components )
```

Comment:
Advance will accept any homogeneous collection of a subclass of GameComponent. Now, logic specific to a game implementation can sit between the GameComponent class and its subclasses in the inheritance hierarchy. The engine can now be free of any reference to Tetris, so it is ripe for use in other applications.